document-store/src/main/java/org/hypertrace/core/documentstore/postgres/PostgresCollection.java

```diff
@@ -500,7 +503,11 @@ public PostgresResultIterator(ResultSet resultSet) {
     @Override
     public boolean hasNext() {
       try {
-        return resultSet.next();
+        if (!didNext) {
+          hasNext = resultSet.next();
+          didNext = true;
```

Comment:
why is the `didNext` state changing on the hasNext call?
I assume that the `didNext` var implies if `next` has been called on the current iterator pointer or not?

Reply:
There is no support for `hastNext` in the underlying `ResultSet` interface. They do have support for functions like `isLast` for traversal and position checking, but some of them are costly.  Secondly, it has a directional concept on the cursor - e.g `TYPE_FORWARD_ONLY`. So, I consider implementing `hashNext` and `next` using `ResultSet's next` only.

As, hashNext has to say that there are more items, but to answer, it advances a cursor, I have to maintain the state.